pkg/admission/kubequota/kubequota_clusterworkspace_monitor.go

```diff
@@ -0,0 +1,125 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubequota
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	tenancyinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/tenancy/v1alpha1"
+)
+
+const clusterWorkspaceDeletionMonitorControllerName = "kcp-kubequota-cluster-workspace-deletion-monitor"
+
+// clusterWorkspaceDeletionMonitor monitors ClusterWorkspaces and terminates QuotaAdmission for a logical cluster
+// when its corresponding ClusterWorkspace is deleted.
+type clusterWorkspaceDeletionMonitor struct {
+	queue    workqueue.RateLimitingInterface
+	stopFunc func(name logicalcluster.Name)
+}
+
+func newClusterWorkspaceDeletionMonitor(
+	clusterWorkspaceInformer tenancyinformers.ClusterWorkspaceInformer,
+	stopFunc func(logicalcluster.Name),
+) *clusterWorkspaceDeletionMonitor {
+	m := &clusterWorkspaceDeletionMonitor{
+		queue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), clusterWorkspaceDeletionMonitorControllerName),
+		stopFunc: stopFunc,
+	}
+
+	clusterWorkspaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: func(obj interface{}) {
+			m.enqueue(obj)
+		},
+	})
+
+	return m
+}
+
+func (m *clusterWorkspaceDeletionMonitor) enqueue(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	m.queue.Add(key)
+}
+
+func (m *clusterWorkspaceDeletionMonitor) Start(stop <-chan struct{}) {
+	defer runtime.HandleCrash()
+	defer m.queue.ShutDown()
+
+	klog.Infof("Starting %s controller", clusterWorkspaceDeletionMonitorControllerName)
+	defer klog.Infof("Shutting down %s controller", clusterWorkspaceDeletionMonitorControllerName)
+
+	go wait.Until(m.startWorker, time.Second, stop)
+
+	<-stop
+}
+
+func (m *clusterWorkspaceDeletionMonitor) startWorker() {
+	for m.processNextWorkItem() {
+	}
+}
+
+func (m *clusterWorkspaceDeletionMonitor) processNextWorkItem() bool {
+	// Wait until there is a new item in the working queue
+	k, quit := m.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer m.queue.Done(key)
+
+	if err := m.process(key); err != nil {
+		runtime.HandleError(fmt.Errorf("clusterWorkspaceDeletionMonitor failed to sync %q, err: %w", key, err))
+
+		m.queue.AddRateLimited(key)
+
+		return true
+	}
+
+	// Clear rate limiting stats on key
+	m.queue.Forget(key)
+
+	return true
+}
+
+func (m *clusterWorkspaceDeletionMonitor) process(key string) error {
+	// e.g. root:org<separator>ws
```

Comment:
I found we lack the helpers to go in both directions here, might be useful to add to the logicalcluster lib later?

Reply:
Maybe later, but it's highly dependent on the format of the key. I eventually want to "sync" all of this key-related stuff, but we're not there yet.